sdk/cosmos/azure_data_cosmos/src/location_cache.rs

```diff
@@ -0,0 +1,570 @@
+use std::{collections::HashMap, sync::Mutex, time::SystemTime};
+
+use serde::{Deserialize, Serialize};
+
+#[derive(Serialize, Deserialize, Debug)]
+pub struct AccountRegion {
+    #[serde(rename = "databaseAccountEndpoint")]
+    pub endpoint: String,
+    #[serde(rename = "name")]
+    pub region: String,
+}
+
+#[derive(Serialize, Deserialize, Debug)]
+pub struct AccountProperties {
+    pub database_account_endpoint: String,
+    pub read_regions: Vec<AccountRegion>,
+    pub write_regions: Vec<AccountRegion>,
+    pub enable_multiple_write_locations: bool,
+}
+
+#[derive(Clone, Copy, PartialEq, Debug, Serialize, Deserialize)]
+pub enum RequestOperation {
+    Read,
+    Write,
+    All,
+}
+
+#[derive(Clone, Default)]
+pub struct DatabaseAccountLocationsInfo {
+    pub preferred_locations: Vec<String>,
+    available_write_locations: Vec<String>,
+    available_read_locations: Vec<String>,
+    available_write_endpoints_by_location: HashMap<String, String>,
+    available_read_endpoints_by_location: HashMap<String, String>,
+    write_endpoints: Vec<String>,
+    read_endpoints: Vec<String>,
+}
+
+#[derive(Serialize, Deserialize, Debug)]
+pub struct LocationUnavailabilityInfo {
+    pub last_check_time: SystemTime,
+    pub unavailable_operation: RequestOperation,
+}
+
+pub struct LocationCache {
+    pub default_endpoint: String,
+    pub locations_info: DatabaseAccountLocationsInfo,
+    pub location_unavailability_info_map: Mutex<HashMap<String, LocationUnavailabilityInfo>>,
```

Comment:
We should use an `RwLock` instead of a `Mutex` here. I'll chat about the difference in our 1:1 :)